In [1]:
!pip install -r requirements.txt

In [2]:
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd 

In [3]:
raw_data= pd.read_csv("data//breast_cancer_dataset_dse.csv")
raw_data.head()

CENTRO     V1  V2  V3  V4  V5   V6     V7  V8  V9  ...        NUT81  \
0       1  10001   1   4   2   2  491  -9532  57   2  ...  5648.296782   
1       1  10001   2  12   2   4  492 -14138  70   2  ...  2821.803323   
2       1  10002   1  10   2  10  891 -10885  61   4  ...  5182.036804   
3       1  10002   2  12   2   2  592  -5872  48   2  ...  7525.644453   
4       1  10003   1  10   2   2  891  -7136  51   2  ...  3333.196596   

         NUT82        NUT83         NUT84         NUT85        NUT86  \
0  3211.984111  5293.483394   9693.117448  24592.425850  4435.469432   
1  1633.620478  2759.095141   5193.159895  10914.074588  2170.876027   
2  2768.285972  4673.497237   8435.682215  18817.064964  4027.165050   
3  4443.970135  7546.378739  13406.852180  30814.012101  5692.820114   
4  1766.302257  3108.772785   5701.329841  14685.569279  2684.650272   

          NUT87        NUT88       NUT89      NUT90  
0   8336.245970  5491.056427   92.785579  71.944429  
1   4026.198830  2813.115847   45.196500  33.327571  
2   5520.428967  4191.977318   72.617601  57.553526  
3  12382.319142  7979.416594  118.245746  91.127254  
4   4775.958162  3194.035958   65.571295  50.351497  

[5 rows x 501 columns]

- v23 (dati non reliable, da considerare per eliminarne? boh)
- v24
- farm 40 
- v6 v7 --- altrimenti togleire v8 e usare v6 e v7 per calcolare età di v17 (età diagnosi)
- v9 a v15 vedere utilità se no togliere 
- v16 da valutare (main diagnosis)
- fum8- fum17 da valutare 
- fis1 - fis8 (lack data in b/ween)
- antr5 
- antr8 - antr11 e antr13 
- da cond1 a open8 -> set a parte > diete 
- alc resto da togliere
- nut50-nut52 e nut55-nut60 mancano i nomi delle variabili 


*da fare*
- v8-fum2
- fum9-fum8
- alc4+alc8+alc12+alc16+alc20+alc25
- medial history da valutare statistically 
- da fam 1 a 32 capire come gestirle (o raggrupparle, o creare 4 variabili vedi chatgpt o valutare come)


In [4]:
alcs_keep = {"ALC4", "ALC8", "ALC12", "ALC16", "ALC20", "ALC25"}
diets_col= ["COND1", "COND2", "COND3", "COND4", "COND5", "COND6", "COND7", "COND8", "COND9", "COND10", "COND11", "COND12",
    "COND13", "COND14", "COND15", "COND16", "COND17", "PASTI1", "PASTI2", "PASTI3", "PASTI4", "PASTI5",
    "ALI1", "ALI2", "ALI3", "ALI4", "ALI5", "ALI6", "ALI7", "ALI8", "ALI9", "ALI10", "ALI11", "CODOP1", "OPEN1",
    "ALI12", "ALI13", "ALI14", "ALI15", "ALI16", "PORZ17", "ALI17", "PORZ18", "ALI18", "PORZ19", "ALI19", "PORZ20",
    "ALI20", "PORZ21", "ALI21", "PORZ22", "ALI22", "PORZ23", "ALI23", "PORZ24", "ALI24", "CODOP2", "POROP2", "OPEN2",
    "ALI25", "PORIAS1", "RIAS1", "PORIAS2", "RIAS2", "ALI26", "ALI27", "PORZ28", "ALI28", "PORZ29", "ALI29", "PORZ30",
    "ALI30", "PORZ31", "ALI31", "PORZ32", "ALI32", "PORZ33", "ALI33", "PORZ34", "ALI34", "PORZ35", "ALI35", "PORZ36",
    "ALI36", "PORZ37", "ALI37", "PORZ38", "ALI38", "PORZ39", "ALI39", "PORZ40", "ALI40", "PORZ41", "ALI41", "PORIAS4",
    "RIAS4", "PORZ42", "ALI42", "PORZ43", "ALI43", "PORZ44", "ALI44", "CODOP3", "POROP3", "OPEN3", "PORIAS3", "RIAS3",
    "PORZ45", "ALI45", "PORZ46", "ALI46", "PORZ47", "ALI47", "PORZ48", "ALI48", "PORZ49", "ALI49", "PORZ50", "ALI50",
    "PORZ51", "ALI51", "ALI52", "MESI52", "PORZ53", "ALI53", "MESI53", "PORZ54", "ALI54", "MESI54", "PORZ55", "ALI55",
    "MESI55", "PORZ56", "ALI56", "MESI56", "PORZ57", "ALI57", "MESI57", "CODOP4", "POROP4", "OPEN4", "MESOP4",
    "CODOP5", "POROP5", "OPEN5", "MESOP5", "PORIAS5", "RIAS5", "PORIAS6", "RIAS6", "CODREG1", "POREG1", "REG1",
    "CODREG2", "POREG2", "REG2", "ALI58", "ALI59", "ALI60", "ALI61", "ALI62", "ALI63", "ALI64", "ALI65", "MESI65",
    "ALI66", "MESI66", "ALI67", "MESI67", "ALI68", "MESI68", "ALI69", "MESI69", "CODOP6", "OPEN6", "MESOP6",
    "CODOP7", "OPEN7", "MESOP7", "RIAS7", "ALI70", "ALI71", "ALI72", "PORZ73", "ALI73", "PORZ74", "ALI74",
    "ALI75", "ALI76", "ALI77", "ALI78", "MESI78", "CODOP8", "POROP8", "OPEN8"
]

cols_to_drop = [f"ALC{i}" for i in range(1, 28) if f"ALC{i}" not in alcs_keep] + [
    "V24", "FARM40", "FIS1", "FIS2", "FIS3", "FIS4", "FIS5", "FIS6", "FIS7", "FIS8",
    "ANTR5", "ANTR8", "ANTR9", "ANTR10", "ANTR11", "ANTR13", "NUT50", "NUT51","NUT51","NUT55",
    "NUT56","NUT57","NUT58","NUT59","NUT60",
]

data_df = raw_data
#.drop(cols_to_drop, axis=1, errors="ignore")

PRESENTAZIONE: Ricordiamoci di spiegare come mai queste variabili sono state ignorate

In [5]:
data_df

CENTRO     V1  V2  V3  V4  V5   V6     V7  V8  V9  ...        NUT81  \
0          1  10001   1   4   2   2  491  -9532  57   2  ...  5648.296782   
1          1  10001   2  12   2   4  492 -14138  70   2  ...  2821.803323   
2          1  10002   1  10   2  10  891 -10885  61   4  ...  5182.036804   
3          1  10002   2  12   2   2  592  -5872  48   2  ...  7525.644453   
4          1  10003   1  10   2   2  891  -7136  51   2  ...  3333.196596   
...      ...    ...  ..  ..  ..  ..  ...    ...  ..  ..  ...          ...   
5152       7  70086   2  14   2  12  294 -14640  74   4  ...  4809.116436   
5153       7  70087   2  14   2  12  294 -12771  69   2  ...  6864.058656   
5154       7  70088   2  14   2  12  294 -13424  70   4  ...  2919.270469   
5155       7  70089   2  14   2   4  294 -13679  71   4  ...  1897.566155   
5156       7  70090   2  14   2   4  294  -9451  60   4  ...  2397.580971   

            NUT82        NUT83         NUT84         NUT85        NUT86  \
0     3211.984111  5293.483394   9693.117448  24592.425850  4435.469432   
1     1633.620478  2759.095141   5193.159895  10914.074588  2170.876027   
2     2768.285972  4673.497237   8435.682215  18817.064964  4027.165050   
3     4443.970135  7546.378739  13406.852180  30814.012101  5692.820114   
4     1766.302257  3108.772785   5701.329841  14685.569279  2684.650272   
...           ...          ...           ...           ...          ...   
5152  2645.048996  4463.223475   8669.116463  21642.053156  3760.635672   
5153  3825.790086  6523.646366  11722.261381  35078.468430  5858.961632   
5154  1554.444277  2697.311293   5105.081750  10997.028813  2268.770773   
5155  1150.334504  1805.547359   3603.283269   8955.762793  1446.452641   
5156  1422.202737  2241.757577   4217.352129  14426.697302  1938.286904   

             NUT87        NUT88       NUT89      NUT90  
0      8336.245970  5491.056427   92.785579  71.944429  
1      4026.198830  2813.115847   45.196500  33.327571  
2      5520.428967  4191.977318   72.617601  57.553526  
3     12382.319142  7979.416594  118.245746  91.127254  
4      4775.958162  3194.035958   65.571295  50.351497  
...            ...          ...         ...        ...  
5152   6939.169466  4661.511688   65.959292  50.958445  
5153  11196.591061  6761.244768  112.616708  86.627207  
5154   3549.838482  2541.161477   47.411014  36.361197  
5155   3319.249391  2129.870453   36.938776  26.708642  
5156   5011.188060  2836.018088   32.908780  26.061385  

[5157 rows x 501 columns]

In [6]:
# Find the columns with missing values and show only the ones with missing values
missing_values = data_df.isnull().sum()
missing_values = missing_values[missing_values > 0]
missing_values

ALC3    1
dtype: int64

Apparently every variable has its own missing value index. 

Notes:

I used this for the null.json file:
https://claude.ai/public/artifacts/aabdaafc-d38e-4dbe-a8fa-4fde253d3e7a

In [7]:
import json
import pandas as pd

with open("data/null.json") as f:
    data = json.load(f)

null_df = pd.DataFrame.from_dict(data, orient="index")
null_df.reset_index(inplace=True)
null_df['index'] = null_df['index'].str.upper()

In [8]:
import json

with open("data/null.json") as f:
    data = json.load(f)

null_df = pd.DataFrame.from_dict(data, orient="index")
null_df.reset_index(inplace=True)
null_df.columns = ["index", "value1", "value2", "value3"][:len(null_df.columns)]
null_df["index"] = null_df["index"].str.upper()
# Replace string "None" with np.nan in value1 and value2
null_df['value1'] = null_df['value1'].replace("None", np.nan)
null_df['value2'] = null_df['value2'].replace("None", np.nan)
null_df['value3'] = null_df['value3'].replace("None", np.nan)

# Convert value1 and value2 columns to numeric types, coercing errors to NaN
null_df['value1'] = pd.to_numeric(null_df['value1'], errors='coerce')
null_df['value2'] = pd.to_numeric(null_df['value2'], errors='coerce')
null_df['value3'] = pd.to_numeric(null_df['value3'], errors='coerce')
null_df = null_df[~null_df["index"].isin(cols_to_drop)].reset_index(drop=True)
null_df

index  value1  value2  value3
0                          V3      99     0.0     NaN
1                          V5      99     0.0     NaN
2                          V6    9999     0.0     NaN
3                          V7       0     NaN     NaN
4                          V8      99     NaN     NaN
..                        ...     ...     ...     ...
375                   RECENZ1      99     NaN     NaN
376                   RECENZ2      99     NaN     NaN
377  MIN(OF RECENZ3-RECENZ7);      99     NaN     NaN
378                   RECENZ8      99     NaN     NaN
379                       V23       9     0.0     NaN

[380 rows x 4 columns]

In [9]:
def replace_with_nulls(data_df, null_df):
    """
    Replace specified values in data_df with NaN based on null_df.
    """
    result_df = data_df.copy()

    for _, row in null_df.iterrows():
        col = row.get("variable", row.get("index"))
        if col not in result_df.columns:
            print(f"Warning: Column '{col}' not in data_df. Skipping.")
            continue

        # Get dtype of the column
        col_dtype = result_df[col].dtype

        # Prepare values to replace with correct type
        vals_to_replace = []
        for key in ["value1", "value2", "value3"]:
            val = row.get(key)
            if pd.notna(val):
                try:
                    val = col_dtype.type(val)  # cast to same type as column
                    vals_to_replace.append(val)
                except Exception as e:
                    print(f"Could not cast value {val} for column {col}: {e}")

        if vals_to_replace:
            result_df[col] = result_df[col].replace(vals_to_replace, pd.NA)

    return result_df

data_df_na = replace_with_nulls(data_df, null_df)

In [10]:
# Check for missing columns in the null_df
null_vars_lower = set(null_df['index'].str.lower())
data_cols_lower = set(map(str.lower, data_df.columns))
cols_to_exclude_lower = set(col.lower() for col in cols_to_drop)

# Step 2: Find variables present in data_df but not mentioned in null_df or excluded
unexpected_columns = list(data_cols_lower - null_vars_lower - cols_to_exclude_lower)

# Optional: Print or inspect the result
print(unexpected_columns)

['nut87', 'nut35', 'nut9', 'fam18', 'nut29', 'nut30', 'nut26', 'nut72', 'nut24', 'nut71', 'fam3', 'nut85', 'centro', 'nut17', 'nut47', 'nut34', 'nut16', 'nut90', 'nut80', 'nut48', 'nut62', 'nut1', 'nut83', 'nut25', 'nut42', 'v2', 'nut6', 'nut79', 'nut2', 'nut88', 'nut69', 'nut18', 'nut4', 'nut43', 'nut31', 'nut82', 'nut65', 'fam23', 'v1', 'nut15', 'fam28', 'fam13', 'nut19', 'nut75', 'nut89', 'nut33', 'nut21', 'nut52', 'v4', 'nut68', 'nut53', 'nut14', 'nut8', 'nut32', 'nut70', 'nut39', 'nut64', 'nut7', 'nut5', 'nut76', 'nut67', 'nut44', 'nut23', 'nut49', 'nut40', 'nut27', 'nut54', 'nut10', 'nut28', 'nut46', 'nut3', 'nut78', 'nut11', 'nut41', 'nut61', 'nut84', 'nut66', 'nut86', 'nut63', 'nut13', 'nut12', 'nut81', 'porp4', 'nut45', 'nut22', 'nut36', 'nut77', 'nut37', 'nut73', 'nut20', 'nut74', 'nut38', 'fam8']


All the variable missing are the ones that do not have a "null" place holder

In [11]:
# Find variables in null_df not present in data_df or cols_to_drop (case-insensitive)
null_vars_lower = set(null_df['index'].str.lower())
data_cols_lower = set(map(str.lower, data_df.columns))
cols_to_exclude_lower = set(col.lower() for col in cols_to_drop)

missing_columns = list(null_vars_lower - data_cols_lower - cols_to_exclude_lower)
print(missing_columns)

['dura8', 'alc4+alc8+alc12+alc16+alc20+alc25', 'porop4', 'etainiz8', 'etainiz1', 'dura1', 'etainiz2', 'v8-fum2', 'porz44', 'dura2', 'fum9-fum8', 'recenz8', 'recenz2', 'min(of etainiz3-etainiz7);', 'min(of recenz3-recenz7);', 'recenz1', 'dura3+dura4+dura5+dura6+dura7']


Check these variables manually, to understand what they are:

- *min(of recenz3-recenz7)*: not in the dataset, neither recenz3–recenz7. Not sure it is a typo.  
- *alc4+alc8+alc12+alc16+alc20+alc25*: total alcohol intake. It is not in the dataframe, but consider creating the variable summing `alc4 + alc8 + alc12 + alc16 + alc20 + alc25`.  
- *v8-fum2*: not in the dataset, but consider creating the variable that counts the years since the smoking cessation.  
- *dura8*: not in the dataset.  
- *recenz8*: not in the dataset.  
- *dura2*: not in the dataset.  
- *recenz2*: not in the dataset.  
- *etainiz1*: not in the dataset.  
- *dura1*: not in the dataset.  
- *dura3+dura4+dura5+dura6+dura7*: not in the dataset.  
- *etainiz8*: not in the dataset.  
- *recenz1*: not in the dataset.  
- *min(of etainiz3-etainiz7)*: not in the dataset.  
- *etainiz2*: not in the dataset.  
- *fum9-fum8*: not in the dataset, but consider creating this variable, which represents the number of years someone is smoking type 1 cigarettes.

In [12]:
data_df_na

CENTRO     V1  V2  V3  V4  V5   V6     V7  V8  V9  ...        NUT81  \
0          1  10001   1   4   2   2  491  -9532  57   2  ...  5648.296782   
1          1  10001   2  12   2   4  492 -14138  70   2  ...  2821.803323   
2          1  10002   1  10   2  10  891 -10885  61   4  ...  5182.036804   
3          1  10002   2  12   2   2  592  -5872  48   2  ...  7525.644453   
4          1  10003   1  10   2   2  891  -7136  51   2  ...  3333.196596   
...      ...    ...  ..  ..  ..  ..  ...    ...  ..  ..  ...          ...   
5152       7  70086   2  14   2  12  294 -14640  74   4  ...  4809.116436   
5153       7  70087   2  14   2  12  294 -12771  69   2  ...  6864.058656   
5154       7  70088   2  14   2  12  294 -13424  70   4  ...  2919.270469   
5155       7  70089   2  14   2   4  294 -13679  71   4  ...  1897.566155   
5156       7  70090   2  14   2   4  294  -9451  60   4  ...  2397.580971   

            NUT82        NUT83         NUT84         NUT85        NUT86  \
0     3211.984111  5293.483394   9693.117448  24592.425850  4435.469432   
1     1633.620478  2759.095141   5193.159895  10914.074588  2170.876027   
2     2768.285972  4673.497237   8435.682215  18817.064964  4027.165050   
3     4443.970135  7546.378739  13406.852180  30814.012101  5692.820114   
4     1766.302257  3108.772785   5701.329841  14685.569279  2684.650272   
...           ...          ...           ...           ...          ...   
5152  2645.048996  4463.223475   8669.116463  21642.053156  3760.635672   
5153  3825.790086  6523.646366  11722.261381  35078.468430  5858.961632   
5154  1554.444277  2697.311293   5105.081750  10997.028813  2268.770773   
5155  1150.334504  1805.547359   3603.283269   8955.762793  1446.452641   
5156  1422.202737  2241.757577   4217.352129  14426.697302  1938.286904   

             NUT87        NUT88       NUT89      NUT90  
0      8336.245970  5491.056427   92.785579  71.944429  
1      4026.198830  2813.115847   45.196500  33.327571  
2      5520.428967  4191.977318   72.617601  57.553526  
3     12382.319142  7979.416594  118.245746  91.127254  
4      4775.958162  3194.035958   65.571295  50.351497  
...            ...          ...         ...        ...  
5152   6939.169466  4661.511688   65.959292  50.958445  
5153  11196.591061  6761.244768  112.616708  86.627207  
5154   3549.838482  2541.161477   47.411014  36.361197  
5155   3319.249391  2129.870453   36.938776  26.708642  
5156   5011.188060  2836.018088   32.908780  26.061385  

[5157 rows x 501 columns]

In [13]:
missing_values = data_df_na.isnull().sum()
missing_values = missing_values[missing_values > 0].sort_values(ascending=False)
missing_values

POREG2    5157
FAM29     5155
FAM24     5149
FAM19     5118
POROP3    5096
          ... 
FAM32        1
GIN12        1
FARM9        1
FARM19       1
FARM28       1
Length: 236, dtype: int64

wAll *FAM_N* are variables pointing out if a relative (not sure which one) is *alive* or *deceased*. They are very high in terms of null values. so I'll discard them.

*GIN28* is the cause of infertility. It is mostly unknown, hence I would discard.

*GIN6* is the type of menopause. It is mostly unknown, hence I would discard.

*FUM_N* is related to information about the smoking amount. Probably we should consider keeping it somehow, but I don't see it happening. We still have information about the smoking age range.**

I would keep the others.
Additional variable to consider removing:
| Code    | Value |
|---------|-------|
| POREG2  | 5157  |
| FAM29   | 5155  |
| FAM24   | 5149  |
| FAM19   | 5118  |
| POROP3  | 5096  |
| POROP8  | 5023  |
| GIN28   | 5015  |
| FAM14   | 4990  |
| COND2   | 4969  |
| POROP2  | 4917  |
| COND8   | 4874  |
| POROP5  | 4857  |
| FUM16   | 4852  |
| POREG1  | 4797  |
| COND4   | 4738  |
| FAM9    | 4467  |
| COND10  | 4357  |
| COND6   | 4354  |
| FUM13   | 4327  |
| FUM17   | 4058  |
| PORZ21  | 3871  |
| FUM10   | 3695  |
| PORZ35  | 3632  |
| PORZ40  | 3611  |
| PORZ73  | 3020  |
| PORZ34  | 3003  |
| PORZ74  | 2964  |
| PORZ48  | 2915  |
| PORZ31  | 2910  |
| PORZ33  | 2802  |
| FAM4    | 2778  |
| PORZ49  | 2628  |
| PORZ22  | 2530  |
| PORZ38  | 2339  |
| PORZ47  | 2335  |
| PORZ56  | 2255  |
| PORZ51  | 2199  |
| PORZ41  | 2118  |
| PORZ20  | 2080  |
| PORZ28  | 2047  |
| PORZ18  | 1958  |
| GIN6    | 1842  |
| PORZ37  | 1786  |
| PORZ23  | 1489  |
| PORZ17  | 1474  |
| PORZ32  | 1422  |
| PORZ53  | 1397  |
| PORZ45  | 1386  |
| PORZ43  | 1363  |
| PORZ42  | 1298  |
| PORZ29  | 1073  |
| PORZ36  | 1040  |
| PORZ50  | 1032  |
| PORZ39  | 985   |





In [14]:
cols_to_remove = [
    'FAM29', 'FAM24', 'FAM19', 'GIN28', 'FAM14', 'FUM16',
    'FAM9', 'FUM13', 'FUM17', 'FUM10', 'FAM4', 'GIN6'
]


cols_to_remove2 = [
    'FAM29', 'FAM24', 'FAM19', 'GIN28', 'FAM14', 'FUM16',
    'FAM9', 'FUM13', 'FUM17', 'FUM10', 'FAM4', 'GIN6',
    'POREG2', 'POROP3', 'POROP8', 'COND2', 'POROP2', 'COND8',
    'POROP5', 'POREG1', 'COND4', 'COND10', 'COND6', 'PORZ21',
    'PORZ35', 'PORZ40', 'PORZ73', 'PORZ34', 'PORZ74', 'PORZ48',
    'PORZ31', 'PORZ33', 'PORZ49', 'PORZ22', 'PORZ38', 'PORZ47',
    'PORZ56', 'PORZ51', 'PORZ41', 'PORZ20', 'PORZ28', 'PORZ18',
    'PORZ37', 'PORZ23', 'PORZ17', 'PORZ32', 'PORZ53', 'PORZ45',
    'PORZ43', 'PORZ42', 'PORZ29', 'PORZ36', 'PORZ50', 'PORZ39'
]

#data_df_na = data_df_na.drop(columns=cols_to_remove, errors='ignore')

Manage "Occasional":

In [15]:
with open("data/occasional.json") as f:
    data = json.load(f)

occasional_df = pd.DataFrame.from_dict(data, orient="index")
occasional_df.reset_index(inplace=True)

In [16]:
occasional_df

index   0
0                             fum3  98
1                             fum4  98
2                             fum5  98
3                           pasti3  98
4                           pasti4  98
..                             ...  ..
134                          alc27  98
135                          dura1  98
136                          dura2  98
137  dura3+dura4+dura5+dura6+dura7  98
138                          dura8  98

[139 rows x 2 columns]

In [17]:

occasional_df['index'] = occasional_df['index'].str.upper()
occasional_df.columns = ["index", "value1"]

new_cols = {}
for _, row in occasional_df.iterrows():
    col = row['index']
    val = row['value1']
    if col in data_df_na.columns:
        new_col_name = f'OCCASIONAL_{col}'
        new_cols[new_col_name] = (data_df_na[col] == val).astype(int)
        data_df_na.loc[data_df_na[col] == val, col] = np.nan

if new_cols:
    data_df_na = pd.concat([data_df_na, pd.DataFrame(new_cols, index=data_df_na.index)], axis=1)
    # Optionally, defragment the DataFrame
    data_df_na = data_df_na.copy()

In [18]:
data_df_na

CENTRO     V1  V2  V3  V4  V5   V6     V7  V8  V9  ... OCCASIONAL_ALC14  \
0          1  10001   1   4   2   2  491  -9532  57   2  ...                0   
1          1  10001   2  12   2   4  492 -14138  70   2  ...                0   
2          1  10002   1  10   2  10  891 -10885  61   4  ...                0   
3          1  10002   2  12   2   2  592  -5872  48   2  ...                0   
4          1  10003   1  10   2   2  891  -7136  51   2  ...                0   
...      ...    ...  ..  ..  ..  ..  ...    ...  ..  ..  ...              ...   
5152       7  70086   2  14   2  12  294 -14640  74   4  ...                0   
5153       7  70087   2  14   2  12  294 -12771  69   2  ...                0   
5154       7  70088   2  14   2  12  294 -13424  70   4  ...                0   
5155       7  70089   2  14   2   4  294 -13679  71   4  ...                0   
5156       7  70090   2  14   2   4  294  -9451  60   4  ...                0   

      OCCASIONAL_ALC16 OCCASIONAL_ALC17 OCCASIONAL_ALC18 OCCASIONAL_ALC20  \
0                    0                0                0                0   
1                    0                0                0                0   
2                    0                0                0                0   
3                    0                0                0                0   
4                    0                0                0                0   
...                ...              ...              ...              ...   
5152                 0                0                0                0   
5153                 0                0                0                0   
5154                 0                0                0                0   
5155                 0                0                0                0   
5156                 0                0                0                0   

     OCCASIONAL_ALC21  OCCASIONAL_ALC22 OCCASIONAL_ALC25 OCCASIONAL_ALC26  \
0                   0                 0                0                0   
1                   0                 0                0                0   
2                   0                 0                0                0   
3                   0                 0                0                0   
4                   0                 0                0                0   
...               ...               ...              ...              ...   
5152                0                 0                0                0   
5153                0                 0                0                0   
5154                0                 0                0                0   
5155                0                 0                0                0   
5156                0                 0                0                0   

     OCCASIONAL_ALC27  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
...               ...  
5152                0  
5153                0  
5154                0  
5155                0  
5156                0  

[5157 rows x 636 columns]

In [19]:
occasional_counts = data_df_na.filter(like='OCCASIONAL_').sum()
print(occasional_counts)

OCCASIONAL_FUM3      0
OCCASIONAL_FUM4      0
OCCASIONAL_FUM5      0
OCCASIONAL_PASTI3    0
OCCASIONAL_PASTI4    0
                    ..
OCCASIONAL_ALC21     0
OCCASIONAL_ALC22     0
OCCASIONAL_ALC25     0
OCCASIONAL_ALC26     0
OCCASIONAL_ALC27     0
Length: 135, dtype: int64


In [20]:
cols_to_drop = [col for col in data_df_na.columns if col.startswith('OCCASIONAL_')]
data_df_na = data_df_na.drop(columns=cols_to_drop)

No occasional variables have positive values. I will drop them all:

### Encode the categorical variables

Note: fam6, fam21, fam26, fam31 have too many unique values to just be encoded. I will treat them separatelly.

In [21]:
data_df_na['V4'] = data_df_na['V4'].replace({1: 0, 2: 1})
data_df_na['V2'] = data_df_na['V2'].replace({2: 0})

In [22]:
data_df_na

CENTRO     V1  V2  V3  V4  V5   V6     V7  V8  V9  ...        NUT81  \
0          1  10001   1   4   1   2  491  -9532  57   2  ...  5648.296782   
1          1  10001   0  12   1   4  492 -14138  70   2  ...  2821.803323   
2          1  10002   1  10   1  10  891 -10885  61   4  ...  5182.036804   
3          1  10002   0  12   1   2  592  -5872  48   2  ...  7525.644453   
4          1  10003   1  10   1   2  891  -7136  51   2  ...  3333.196596   
...      ...    ...  ..  ..  ..  ..  ...    ...  ..  ..  ...          ...   
5152       7  70086   0  14   1  12  294 -14640  74   4  ...  4809.116436   
5153       7  70087   0  14   1  12  294 -12771  69   2  ...  6864.058656   
5154       7  70088   0  14   1  12  294 -13424  70   4  ...  2919.270469   
5155       7  70089   0  14   1   4  294 -13679  71   4  ...  1897.566155   
5156       7  70090   0  14   1   4  294  -9451  60   4  ...  2397.580971   

            NUT82        NUT83         NUT84         NUT85        NUT86  \
0     3211.984111  5293.483394   9693.117448  24592.425850  4435.469432   
1     1633.620478  2759.095141   5193.159895  10914.074588  2170.876027   
2     2768.285972  4673.497237   8435.682215  18817.064964  4027.165050   
3     4443.970135  7546.378739  13406.852180  30814.012101  5692.820114   
4     1766.302257  3108.772785   5701.329841  14685.569279  2684.650272   
...           ...          ...           ...           ...          ...   
5152  2645.048996  4463.223475   8669.116463  21642.053156  3760.635672   
5153  3825.790086  6523.646366  11722.261381  35078.468430  5858.961632   
5154  1554.444277  2697.311293   5105.081750  10997.028813  2268.770773   
5155  1150.334504  1805.547359   3603.283269   8955.762793  1446.452641   
5156  1422.202737  2241.757577   4217.352129  14426.697302  1938.286904   

             NUT87        NUT88       NUT89      NUT90  
0      8336.245970  5491.056427   92.785579  71.944429  
1      4026.198830  2813.115847   45.196500  33.327571  
2      5520.428967  4191.977318   72.617601  57.553526  
3     12382.319142  7979.416594  118.245746  91.127254  
4      4775.958162  3194.035958   65.571295  50.351497  
...            ...          ...         ...        ...  
5152   6939.169466  4661.511688   65.959292  50.958445  
5153  11196.591061  6761.244768  112.616708  86.627207  
5154   3549.838482  2541.161477   47.411014  36.361197  
5155   3319.249391  2129.870453   36.938776  26.708642  
5156   5011.188060  2836.018088   32.908780  26.061385  

[5157 rows x 501 columns]

In [23]:
with open('data//to_be_encoded.txt', 'r') as f:
    file_content = f.read().strip()
    cols_to_encode_raw = [col.strip().upper() for col in file_content.split(';') if col.strip()]

# Keep only columns that exist in the dataframe
cols_to_encode = [col for col in cols_to_encode_raw if col in data_df_na.columns]

# Report any missing columns
missing_cols = [col for col in cols_to_encode_raw if col not in data_df_na.columns]
if missing_cols:
    print(f"Warning: The following columns were not found in the dataframe: {missing_cols}")

# One-hot encode the columns that exist
data_df_encoded = pd.get_dummies(
    data_df_na,
    columns=cols_to_encode,
    prefix=cols_to_encode,
    prefix_sep='_'
)
data_df_encoded

V1  V2  V3   V6     V7  V8 V10  V11 V12   V13  ...  FARM31_0  \
0     10001   1   4  491  -9532  57  20    2   5     5  ...      True   
1     10001   0  12  492 -14138  70  26    2   3     3  ...      True   
2     10002   1  10  891 -10885  61  26    1   6     6  ...      True   
3     10002   0  12  592  -5872  48  24    2   4     5  ...      True   
4     10003   1  10  891  -7136  51  21    2  10    13  ...      True   
...     ...  ..  ..  ...    ...  ..  ..  ...  ..   ...  ...       ...   
5152  70086   0  14  294 -14640  74  19    3   5     5  ...      True   
5153  70087   0  14  294 -12771  69  26    5   7     7  ...      True   
5154  70088   0  14  294 -13424  70  19    1   5  <NA>  ...      True   
5155  70089   0  14  294 -13679  71  20    2   7     5  ...      True   
5156  70090   0  14  294  -9451  60  26    3   5     5  ...      True   

     FARM31_4 FARM35_0 FARM35_4  FARM39_0 FARM39_1  V23_1  V23_2  V23_3  V23_4  
0       False     True    False      True    False  False   True  False  False  
1       False     True    False      True    False  False  False   True  False  
2       False     True    False      True    False  False   True  False  False  
3       False     True    False      True    False  False   True  False  False  
4       False     True    False      True    False  False   True  False  False  
...       ...      ...      ...       ...      ...    ...    ...    ...    ...  
5152    False     True    False      True    False  False  False   True  False  
5153    False     True    False      True    False  False   True  False  False  
5154    False     True    False      True    False  False   True  False  False  
5155    False     True    False      True    False  False   True  False  False  
5156    False     True    False      True    False  False  False   True  False  

[5157 rows x 1221 columns]

740 columns!|

### Variable Creation

In [24]:
data_df_encoded['ALC_TOTAL'] = data_df_encoded[['ALC4', 'ALC8', 'ALC12', 'ALC16', 'ALC20', 'ALC25']].sum(axis=1, skipna=True)
data_df_encoded['FUM_TYPE1_YEARS'] = data_df_encoded[['FUM9', 'FUM8']].assign(FUM8=lambda x: -x['FUM8']).sum(axis=1, skipna=True)
data_df_encoded['PRE_BMI'] = data_df_encoded['ANTR1'] / (data_df_encoded['ANTR2'] / 100) ** 2 # Body Mass Index (BMI) pre diagnosis
data_df_encoded['WHR'] = data_df_encoded['ANTR3'] / data_df_encoded['ANTR4'] # Waist-to-Hip Ratio (WHR)
#data_df_encoded['WEIGHT_MAX_DIFF'] = data_df_encoded['ANTR8'] - data_df_encoded['ANTR10'] # Max weight variation: max - min since 20 years
data_df_encoded['AVG_USUAL_WEIGHT'] = data_df_encoded[['ANTR6', 'ANTR7']].mean(axis=1, skipna=True) # Usual Weight 30-50 years old

C:\Users\maidi\AppData\Local\Temp\ipykernel_20016\674797425.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  data_df_encoded['ALC_TOTAL'] = data_df_encoded[['ALC4', 'ALC8', 'ALC12', 'ALC16', 'ALC20', 'ALC25']].sum(axis=1, skipna=True)
C:\Users\maidi\AppData\Local\Temp\ipykernel_20016\674797425.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  data_df_encoded['FUM_TYPE1_YEARS'] = data_df_encoded[['FUM9', 'FUM8']].assign(FUM8=lambda x: -x['FUM8']).sum(axis=1, skipna=True)
C:\Users\maidi\AppData\Local\Temp\ipykernel_20016\6747

In [25]:
pd.set_option('future.no_silent_downcasting', True)
corr_matrix = data_df_encoded.fillna(np.nan).infer_objects(copy=False).corr().abs()
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))

# Find columns with correlation above threshold (e.g., 0.9)
threshold = 0.9
high_corr = [(col, row, upper.loc[row, col])
             for col in upper.columns
             for row in upper.index
             if upper.loc[row, col] > threshold]

# Convert to DataFrame for easier viewing
high_corr_df = pd.DataFrame(high_corr, columns=["Variable 1", "Variable 2", "Correlation"])

In [26]:
high_corr_df

Variable 1 Variable 2  Correlation
0                  V8         V7     0.996893
1                FUM8       FUM6     0.996048
2                FUM9       FUM6     0.920339
3                FUM9       FUM8     0.922803
4               FUM12      FUM11     0.972749
..                ...        ...          ...
569          FARM39_1   FARM35_0     1.000000
570          FARM39_1   FARM35_4     1.000000
571          FARM39_1   FARM39_0     1.000000
572           PRE_BMI      ANTR1     0.907436
573  AVG_USUAL_WEIGHT      ANTR7     0.948046

[574 rows x 3 columns]

In [27]:
import networkx as nx

G = nx.Graph()
for _, row in high_corr_df.iterrows():
    G.add_edge(row["Variable 1"], row["Variable 2"])

removable_vars = set()
for component in nx.connected_components(G):
    # For each group, count the number of high correlations for each variable
    subgraph = G.subgraph(component)
    degrees = dict(subgraph.degree())
    # Keep the variable with the highest degree (most connections)
    keep_var = max(degrees, key=degrees.get)
    # Mark all others as removable
    removable_vars.update(set(component) - {keep_var})

removable_vars = sorted(removable_vars)
print("Variables to consider removing:", removable_vars)

Variables to consider removing: ['ALC1_1', 'ALC2_1', 'AVG_USUAL_WEIGHT', 'CENTRO_3', 'CODOP4', 'COND12_1', 'COND13_1', 'COND15_1', 'FAM10', 'FAM11_0', 'FAM11_1469', 'FAM11_1986', 'FAM12', 'FAM13_0', 'FAM13_7', 'FAM16_0', 'FAM16_1610', 'FAM16_2273', 'FAM17', 'FAM18_0', 'FAM18_5', 'FAM18_8', 'FAM20', 'FAM21_0', 'FAM21_1729', 'FAM21_1890', 'FAM22', 'FAM23_0', 'FAM23_4', 'FAM23_9', 'FAM24_1', 'FAM24_2', 'FAM25', 'FAM26_0', 'FAM26_1590', 'FAM26_1749', 'FAM26_1889', 'FAM28_0', 'FAM28_3', 'FAM28_5', 'FAM29_2', 'FAM30', 'FAM31_0', 'FAM31_1599', 'FAM31_1749', 'FAM3_0', 'FAM5', 'FAM6_0', 'FARM1', 'FARM11_0', 'FARM13', 'FARM19_0', 'FARM21', 'FARM22', 'FARM23_0', 'FARM23_1', 'FARM23_3', 'FARM25', 'FARM26', 'FARM27_0', 'FARM27_1', 'FARM27_5', 'FARM28', 'FARM29', 'FARM30', 'FARM31_0', 'FARM31_4', 'FARM32', 'FARM33', 'FARM34', 'FARM35_0', 'FARM35_4', 'FARM36', 'FARM37', 'FARM38', 'FARM39_0', 'FARM39_1', 'FARM40', 'FARM7_0', 'FIS6_0', 'FIS8_0', 'FUM11', 'FUM13_7', 'FUM14', 'FUM6', 'FUM9', 'GIN4_3', 'G

PRE_BMI is highly correlated with ANTR1

In [28]:
# Keep V8 and consider removing V7 instead (I prefer to keep Age, since it is easier to interpret)
always_keep = ['V8', 'CENTRO', 'V1', 'V2', 'V3', 'V5', 'V6']

removable_vars = [var for var in removable_vars if var not in always_keep]
if 'V8' in always_keep and 'V7' not in removable_vars:
    removable_vars.append('V7')
removable_vars = [var for var in removable_vars if not var.startswith('NUT')] #We need these variables for the dietary patterns

Should we keep NUT variables, since you will look for diatetary patterns?

In [29]:
filtered_df = data_df_encoded.drop(columns=removable_vars, errors="ignore")

pd.set_option('future.no_silent_downcasting', True)
corr_matrix = filtered_df.fillna(np.nan).infer_objects(copy=False).corr().abs()
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))

threshold = 0.9
high_corr = [
    (col, row, upper.loc[row, col])
    for col in upper.columns
    for row in upper.index
    if upper.loc[row, col] > threshold
]

high_corr_df = pd.DataFrame(high_corr, columns=["Variable 1", "Variable 2", "Correlation"])
high_corr_df

Variable 1 Variable 2  Correlation
0         NUT4       NUT2     0.937045
1        NUT11       NUT3     0.926288
2        NUT23       NUT5     0.953774
3        NUT23       NUT7     0.905227
4        NUT24       NUT6     0.924585
..         ...        ...          ...
249      NUT88      NUT84     0.967073
250      NUT88      NUT85     0.968499
251      NUT88      NUT86     0.927043
252      NUT88      NUT87     0.966944
253      NUT90      NUT89     0.995014

[254 rows x 3 columns]

In [30]:
filtered_df1 = data_df_encoded.drop(columns=removable_vars, errors="ignore")
filtered_df1

V1  V2  V3   V6  V8 V10  V11 V12   V13   V16  ... FARM19_8 FARM23_2  \
0     10001   1   4  491  57  20    2   5     5  1744  ...    False    False   
1     10001   0  12  492  70  26    2   3     3  7071  ...    False    False   
2     10002   1  10  891  61  26    1   6     6  1748  ...    False    False   
3     10002   0  12  592  48  24    2   4     5  4549  ...    False    False   
4     10003   1  10  891  51  21    2  10    13  1743  ...    False    False   
...     ...  ..  ..  ...  ..  ..  ...  ..   ...   ...  ...      ...      ...   
5152  70086   0  14  294  74  19    3   5     5  8232  ...    False    False   
5153  70087   0  14  294  69  26    5   7     7  8232  ...    False    False   
5154  70088   0  14  294  70  19    1   5  <NA>  8220  ...    False    False   
5155  70089   0  14  294  71  20    2   7     5  7242  ...    False    False   
5156  70090   0  14  294  60  26    3   5     5  7234  ...    False    False   

     FARM23_5  V23_1  V23_2  V23_3  V23_4 ALC_TOTAL FUM_TYPE1_YEARS       WHR  
0       False  False   True  False  False    1010.0               0  0.859813  
1       False  False  False   True  False       7.0               0  0.952381  
2       False  False   True  False  False      21.0               0  0.810811  
3       False  False   True  False  False      16.0               0  0.817518  
4       False  False   True  False  False       3.0               0  0.764706  
...       ...    ...    ...    ...    ...       ...             ...       ...  
5152    False  False  False   True  False       7.0               0      <NA>  
5153    False  False   True  False  False    1005.0               0  0.729412  
5154    False  False   True  False  False       7.0               0      <NA>  
5155    False  False   True  False  False       7.0               0  0.842105  
5156    False  False  False   True  False       7.0               0      <NA>  

[5157 rows x 1133 columns]

In [31]:
pd.set_option('future.no_silent_downcasting', True)
corr_matrix = filtered_df1.fillna(np.nan).infer_objects(copy=False).corr().abs()
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))

# Find columns with correlation above threshold (e.g., 0.9)
threshold = 0.80
high_corr = [(col, row, upper.loc[row, col])
             for col in upper.columns
             for row in upper.index
             if upper.loc[row, col] > threshold]

# Convert to DataFrame for easier viewing
high_corr_df = pd.DataFrame(high_corr, columns=["Variable 1", "Variable 2", "Correlation"])

In [32]:
high_corr_df

Variable 1 Variable 2  Correlation
0           V3         V1     0.806242
1          V16         V2     0.855472
2         FUM8       FUM1     0.822120
3        ANTR4      ANTR3     0.835260
4       MESOP4      PORP4     0.809823
..         ...        ...          ...
523    GIN25_0      GIN18     0.878314
524   FARM15_0     FARM16     0.856691
525   FARM19_5   FARM15_5     0.894340
526   FARM19_8     FARM18     0.879296
527   FARM23_2     FARM20     0.819750

[528 rows x 3 columns]

In [33]:
G = nx.Graph()
for _, row in high_corr_df.iterrows():
    G.add_edge(row["Variable 1"], row["Variable 2"])

removable_vars = set()
for component in nx.connected_components(G):
    # For each group, count the number of high correlations for each variable
    subgraph = G.subgraph(component)
    degrees = dict(subgraph.degree())
    # Keep the variable with the highest degree (most connections)
    keep_var = max(degrees, key=degrees.get)
    # Mark all others as removable
    removable_vars.update(set(component) - {keep_var})

removable_vars = sorted(removable_vars)
print("Variables to consider removing:", removable_vars)

Variables to consider removing: ['ALC2_0', 'ALC6', 'ANAM23', 'ANTR3', 'ANTR7', 'CODOP5', 'CODOP7', 'CODOP8', 'COND11_1', 'COND14_2', 'COND17_1', 'COND2_4', 'FAM14_2', 'FAM4_2', 'FARM15_0', 'FARM15_5', 'FARM19_8', 'FARM20', 'FARM8', 'FARM9', 'FIS4_9', 'FIS8_9', 'FUM1', 'GIN10', 'GIN17', 'GIN21_0', 'GIN22_0', 'GIN23_0', 'GIN25_0', 'GIN4_1', 'ICDAN11', 'MESOP4', 'NUT1', 'NUT10', 'NUT11', 'NUT12', 'NUT14', 'NUT15', 'NUT16', 'NUT17', 'NUT18', 'NUT19', 'NUT2', 'NUT22', 'NUT23', 'NUT24', 'NUT25', 'NUT27', 'NUT28', 'NUT3', 'NUT30', 'NUT31', 'NUT32', 'NUT33', 'NUT34', 'NUT35', 'NUT37', 'NUT38', 'NUT39', 'NUT4', 'NUT40', 'NUT41', 'NUT44', 'NUT46', 'NUT5', 'NUT51', 'NUT52', 'NUT53', 'NUT54', 'NUT55', 'NUT6', 'NUT60', 'NUT61', 'NUT62', 'NUT63', 'NUT64', 'NUT65', 'NUT66', 'NUT67', 'NUT68', 'NUT69', 'NUT7', 'NUT70', 'NUT71', 'NUT72', 'NUT73', 'NUT74', 'NUT75', 'NUT76', 'NUT78', 'NUT79', 'NUT8', 'NUT80', 'NUT81', 'NUT82', 'NUT83', 'NUT84', 'NUT85', 'NUT86', 'NUT87', 'NUT88', 'NUT89', 'NUT9', 'NUT90',

In [34]:
# Keep V8 and consider removing V7 instead (I prefer to keep Age, since it is easier to interpret)
always_keep = ['V8', 'CENTRO', 'V1', 'V2', 'V3', 'V5', 'V6'] + diets_col

removable_vars = [var for var in removable_vars if var not in always_keep]
if 'V8' in always_keep and 'V7' not in removable_vars:
    removable_vars.append('V7')
#removable_vars = [var for var in removable_vars if not var.startswith('NUT')] #We need these variables for the dietary patterns

In [35]:
filtered_df2 = filtered_df1.drop(columns=removable_vars, errors="ignore")
filtered_df2

V1  V2  V3   V6  V8 V10  V11 V12   V13  V17  ... FARM19_5 FARM23_2  \
0     10001   1   4  491  57  20    2   5     5  491  ...    False    False   
1     10001   0  12  492  70  26    2   3     3  392  ...    False    False   
2     10002   1  10  891  61  26    1   6     6  391  ...    False    False   
3     10002   0  12  592  48  24    2   4     5  492  ...    False    False   
4     10003   1  10  891  51  21    2  10    13  891  ...    False    False   
...     ...  ..  ..  ...  ..  ..  ...  ..   ...  ...  ...      ...      ...   
5152  70086   0  14  294  74  19    3   5     5  294  ...    False    False   
5153  70087   0  14  294  69  26    5   7     7  194  ...    False    False   
5154  70088   0  14  294  70  19    1   5  <NA>  294  ...    False    False   
5155  70089   0  14  294  71  20    2   7     5  194  ...    False    False   
5156  70090   0  14  294  60  26    3   5     5  294  ...    False    False   

      FARM23_5  V23_1  V23_2  V23_3  V23_4 ALC_TOTAL FUM_TYPE1_YEARS       WHR  
0        False  False   True  False  False    1010.0               0  0.859813  
1        False  False  False   True  False       7.0               0  0.952381  
2        False  False   True  False  False      21.0               0  0.810811  
3        False  False   True  False  False      16.0               0  0.817518  
4        False  False   True  False  False       3.0               0  0.764706  
...        ...    ...    ...    ...    ...       ...             ...       ...  
5152     False  False  False   True  False       7.0               0      <NA>  
5153     False  False   True  False  False    1005.0               0  0.729412  
5154     False  False   True  False  False       7.0               0      <NA>  
5155     False  False   True  False  False       7.0               0  0.842105  
5156     False  False  False   True  False       7.0               0      <NA>  

[5157 rows x 1021 columns]

In [38]:
# Choose your option
w_df = filtered_df1.copy() # filtered_df1.copy() / filtered_df2.copy()
w_df.to_csv("data/allcolumns_90.csv", index=False)